File: advent-of-code/2018/day2/part1.c
#include <stdio.h>
#include <string.h>

int main() {
  int k[] = {0, 0};
  char line[27];
  while (scanf("%26s\n", line) != EOF) {
    char l[26];
    bzero(l,26);
    for (char *p = line; *p; p++) l[*p - 'a']++;
    int s[] = {0, 0};
    for (int i = 0; i < 26; i++) {
      if (l[i] == 2) s[0]++;
      else if (l[i] == 3) s[1]++;
    }
    if (s[0]) k[0]++;
    if (s[1]) k[1]++;
  }
  printf("%d\n", k[0] * k[1]);
}

File: advent-of-code/2021/day13/part1.hs
#!/usr/bin/env runhaskell
import Data.Char
import Data.List
import System.IO

parseNumbers :: String -> Int -> (Int, Int)
parseNumbers (',':cs) acc = (read cs, acc)
parseNumbers (c:cs) acc = parseNumbers cs $ acc * 10 + digitToInt c

parsePoint :: String -> (Int, Int)
parsePoint s = parseNumbers s 0

readPoints :: IO [(Int, Int)]
readPoints = do
  line <- getLine
  if null line
    then return []
    else do
      ps <- readPoints
      return $ parsePoint line : ps

parseCoord :: String -> (Bool, Int)
parseCoord ('x':(_:s)) = (True, read s)
parseCoord ('y':(_:s)) = (False, read s)

parseInstruction :: String -> (Bool, Int)
parseInstruction s = parseCoord $ words s !! 2

readInstructions :: IO [(Bool, Int)]
readInstructions = do
  eof <- isEOF
  if eof
    then return []
    else do
      line <- getLine; is <- readInstructions
      return $ parseInstruction line : is

fold :: [(Int, Int)] -> (Bool, Int) -> [(Int, Int)]
fold [] _ = []
fold (p@(_,x):ps) (i@(True, n)) | x < n = p : fold ps i
fold ((y,x):ps) (i@(True, n)) = (y,2*n-x) : fold ps i
fold (p@(y,_):ps) (i@(_, n)) | y < n = p : fold ps i
fold ((y,x):ps) (i@(_, n)) = (2*n-y,x) : fold ps i

main = do
  ps <- readPoints
  is <- readInstructions
  putStrLn $ show $ length $ nub $ fold ps $ head is
